forch/forchestrator.py

```
@@ -292,22 +293,49 @@ def load_config():
     with open(config_path, 'r') as stream:
         return yaml.safe_load(stream)
 
+
+def show_error(path, params):
+    """Display errors"""
+    return f"Cannot initialize forch: {str(error)}"
+
+
 if __name__ == '__main__':
     logging.basicConfig(level=logging.INFO)
-    CONFIG = load_config()
-    FORCH = Forchestrator(CONFIG)
-    FORCH.initialize()
-    HTTP = forch.http_server.HttpServer(CONFIG.get('http', {}),
-                                        FORCH.get_local_port())
-    HTTP.map_request('system_state', FORCH.get_system_state)
-    HTTP.map_request('dataplane_state', FORCH.get_dataplane_state)
-    HTTP.map_request('switch_state', FORCH.get_switch_state)
-    HTTP.map_request('cpn_state', FORCH.get_cpn_state)
-    HTTP.map_request('process_state', FORCH.get_process_state)
-    HTTP.map_request('host_path', FORCH.get_host_path)
-    HTTP.map_request('list_hosts', FORCH.get_list_hosts)
-    HTTP.map_request('', HTTP.static_file(''))
+    CONFIG = {}
+    FORCH = None
+
+    try:
+        CONFIG = load_config()
+        FORCH = Forchestrator(CONFIG)
+        FORCH.initialize()
+    except Exception as e:
+        error = e
+
+    http_port = FORCH.get_local_port() if FORCH else _DEFAULT_PORT
```

I think the logical problem here is that if forch fails to initialize, then _DEFAULT_PORT isn't really a good backup, since it might be different than what people are expecting, and then still give a 404 because they would be looking at the wrong port. I think we might just have to give up the ghost and fail hard if the configuration is unparseable, otherwise I think the behavior is too mysterious. How about something like:

CONFIG = load_config()
FORCH = Forchestrator(CONFIG)
HTTP = HttpServer()

try:
  FORCH.initialize()
  HTTP.map_request()
  ...
except
  HTTP.map_request('', show_error)

???  That way, we have control over the error handling, since we can put most/all stuff in initialize() -- but still use get_local_port()?

Thoughts?

---

>>> _DEFAULT_PORT isn't really a good backup
Agree

>>> I think we might just have to give up the ghost and fail hard if the configuration is unparseable, otherwise I think the behavior is too mysterious

I will add a error logging if cannot load the config successfully then

>>> That way, we have control over the error handling, since we can put most/all stuff in initialize()

Had the same idea... Will check what should be moved to initialize.